Fix NaN error in dynamic quantization in qLinear, re-enable test_quantized_rnn

raghuramank100 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37159 Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell
* #35961 Fix weight quantization in RNNs
* **#36009 Fix NaN error in dynamic quantization in qLinear, re-enable test_quantized_rnn**

When scale is very small (less than float eps, but greater than minimum double precision value), computation of reciprocal of scale in floating point precision within FBGEMM returns inf, while QuantUtils does not. Changed computation in QuantUtils to occur with floating point precision to re-enable tests.

Differential Revision: [D20853000](https://our.internmc.facebook.com/intern/diff/D20853000/)

raghuramank10000 committed:
Fix NaN error in dynamic quantization in qLinear, re-enable test_quantized_rnn

When scale is very small (less than float eps, but greater than minimum double precision value), computation of reciprocal of scale in floating point precision within FBGEMM returns inf, while QuantUtils does not. Changed computation in QuantUtils to occur with floating point precision to re-enable tests.

Differential Revision: [D20853000](https://our.internmc.facebook.com/intern/diff/D20853000/)

[ghstack-poisoned]

jianyuh commented:
Still see the errors in the OSS unit test:
https://circleci.com/gh/pytorch/pytorch/5053092?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

z-a-f commented:
This change doesn't fix the flakiness of the test. The nan error happens because inside the observer we get `qmin > qmax`, not because of the float. In fact, `float(np.nan)` is a valid NaN value.

z-a-f commented:
I will take a shot at this issue tomorrow evening, if this is not fixed by then

z-a-f reviewed (changes requested):
As mentioned -- this change reenables the test without fixing the flakiness.

supriyar commented on test/quantization/test_quantization.py:
```diff
@@ -575,7 +575,6 @@ def checkQuantized(model):
         checkQuantized(model)
 
-    @unittest.skip("temporarily disable the test")
     @given(qengine=st.sampled_from(("fbgemm",)))
```
@raghuramank100 could you add `qnnpack` here as well? I believe it was skipped for the same reason.

raghuramank10000 committed:
Update on "Fix NaN error in dynamic quantization in qLinear, re-enable test_quantized_rnn"


When scale is very small (less than float eps, but greater than minimum double precision value), computation of reciprocal of scale in floating point precision within FBGEMM returns inf, while QuantUtils does not. Changed computation in QuantUtils to occur with floating point precision to re-enable tests.

Differential Revision: [D20853000](https://our.internmc.facebook.com/intern/diff/D20853000/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fix NaN error in dynamic quantization in qLinear, re-enable test_quantized_rnn"


When scale is very small (less than float eps, but greater than minimum double precision value), computation of reciprocal of scale in floating point precision within FBGEMM returns inf, while QuantUtils does not. Changed computation in QuantUtils to occur with floating point precision to re-enable tests.

Differential Revision: [D20853000](https://our.internmc.facebook.com/intern/diff/D20853000/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fix NaN error in dynamic quantization in qLinear, re-enable test_quantized_rnn"


When scale is very small (less than float eps, but greater than minimum double precision value), computation of reciprocal of scale in floating point precision within FBGEMM returns inf, while QuantUtils does not. Changed computation in QuantUtils to occur with floating point precision to re-enable tests.

Differential Revision: [D20853000](https://our.internmc.facebook.com/intern/diff/D20853000/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fix NaN error in dynamic quantization in qLinear, re-enable test_quantized_rnn"


When scale is very small (less than float eps, but greater than minimum double precision value), computation of reciprocal of scale in floating point precision within FBGEMM returns inf, while QuantUtils does not. Changed computation in QuantUtils to occur with floating point precision to re-enable tests.

Differential Revision: [D20853000](https://our.internmc.facebook.com/intern/diff/D20853000/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fix NaN error in dynamic quantization in qLinear, re-enable test_quantized_rnn"


When scale is very small (less than float eps, but greater than minimum double precision value), computation of reciprocal of scale in floating point precision within FBGEMM returns inf, while QuantUtils does not. Changed computation in QuantUtils to occur with floating point precision to re-enable tests.

Differential Revision: [D20853000](https://our.internmc.facebook.com/intern/diff/D20853000/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fix NaN error in dynamic quantization in qLinear, re-enable test_quantized_rnn"


When scale is very small (less than float eps, but greater than minimum double precision value), computation of reciprocal of scale in floating point precision within FBGEMM returns inf, while QuantUtils does not. Changed computation in QuantUtils to occur with floating point precision to re-enable tests.

Differential Revision: [D20853000](https://our.internmc.facebook.com/intern/diff/D20853000/)

[ghstack-poisoned]

z-a-f commented on test/quantization/test_quantize.py:
```diff
@@ -624,208 +624,195 @@ def checkQuantized(model):
+
+        ref_out, ref_hid = ref(x, hiddens)
+
+        for qengine in ['fbgemm', 'qnnpack']:
```
This will fail because `qengine` is not defined

z-a-f commented on test/quantization/test_quantize.py:
```diff
@@ -624,208 +624,195 @@ def checkQuantized(model):
+
+
+    def test_default_quantized_lstm(self):
+        for qengine in ['fbgemm', 'qnnpack']:
```
Same here

raghuramank10000 committed:
Update on "Fix NaN error in dynamic quantization in qLinear, re-enable test_quantized_rnn"


When scale is very small (less than float eps, but greater than minimum double precision value), computation of reciprocal of scale in floating point precision within FBGEMM returns inf, while QuantUtils does not. Changed computation in QuantUtils to occur with floating point precision to re-enable tests.

Differential Revision: [D20853000](https://our.internmc.facebook.com/intern/diff/D20853000/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fix NaN error in dynamic quantization in qLinear, re-enable test_quantized_rnn"


When scale is very small (less than float eps, but greater than minimum double precision value), computation of reciprocal of scale in floating point precision within FBGEMM returns inf, while QuantUtils does not. Changed computation in QuantUtils to occur with floating point precision to re-enable tests.

Differential Revision: [D20853000](https://our.internmc.facebook.com/intern/diff/D20853000/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fix NaN error in dynamic quantization in qLinear, re-enable test_quantized_rnn"


When scale is very small (less than float eps, but greater than minimum double precision value), computation of reciprocal of scale in floating point precision within FBGEMM returns inf, while QuantUtils does not. Changed computation in QuantUtils to occur with floating point precision to re-enable tests.

Differential Revision: [D20853000](https://our.internmc.facebook.com/intern/diff/D20853000/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fix NaN error in dynamic quantization in qLinear, re-enable test_quantized_rnn"


When scale is very small (less than float eps, but greater than minimum double precision value), computation of reciprocal of scale in floating point precision within FBGEMM returns inf, while QuantUtils does not. Changed computation in QuantUtils to occur with floating point precision to re-enable tests.

Differential Revision: [D20853000](https://our.internmc.facebook.com/intern/diff/D20853000/)

[ghstack-poisoned]

z-a-f commented on test/quantization/test_quantize.py:
```diff
@@ -624,208 +624,201 @@ def checkQuantized(model):
+            if qengine == 'qnnpack':
+                if IS_PPC or TEST_WITH_UBSAN or IS_MACOS or IS_WINDOWS:
+                    continue
+            with override_quantized_engine(qengine):
+                if qengine in torch.backends.quantized.supported_engines:
```
I think this line should be before the `override_quantized_engine`:

```python
if qengine in torch.backends.quantized.supported_engines:
  continue
with override_quantized_engine(qengine):
```

z-a-f commented:
Reverting because of the test_qelu breakage https://app.circleci.com/pipelines/github/pytorch/pytorch/160770/workflows/a3ca7391-3ba4-4558-9929-7ff1ffd4794c/
